packages/contracts/contracts/proof-of-use/groups/UserGroups.sol

```diff
@@ -58,28 +70,33 @@ contract UserGroups is OffChainOwnable {
       "Tried to union with global group"
     );
 
-    UnionFind.Group storage newGroup = UnionFind.unionStruct(groups, groupA, groupB);
+    UnionFind.Group storage newGroup = UnionFind.unionStruct(groupA, groupB);
 
     require(newGroup.size <= maxGroupSize, "Maximum group size exceeded");
   }
 
   function allowGlobalGroupTransfers(address a) external {
-    UnionFind.Group storage globalGroup = UnionFind.findStruct(groups, bytes32(GLOBAL_GROUP));
-    UnionFind.Group storage superGroup = UnionFind.findStruct(groups, bytes32(SUPER_GROUP));
+    UnionFind.Group storage globalGroup = UnionFind.findStruct(GLOBAL_GROUP);
+    UnionFind.Group storage superGroup = UnionFind.findStruct(SUPER_GROUP);
 
-    UnionFind.Group storage group = UnionFind.findStruct(groups, bytes32(a));
+    UnionFind.Group storage group = UnionFind.findStruct(a);
 
     require(group.parent != globalGroup.parent && group.parent != superGroup.parent, "Address a is reserved");
     require(
-      isUseProven(address(group.parent)),
+      isUseProven(group.usedTokens, group.totalTokens),
       "The use hasn't been proven yet"
     );
 
-    UnionFind.union(groups, globalGroup.parent, group.parent);
+    UnionFind.unionStruct(globalGroup, group);
   }
 
-  function isUseProven(address root) internal view returns (bool) {
-    // TODO(ritave): Connect with TCR
-    return true;
+  function isUseProven(uint usedTokens, uint totalTokens) internal view returns (bool) {
```

Comment:
we can't merge this into master until the whitelisting is ready, right? because it would prevent people from applying to TCR via multisig-newsroom

Reply:
Whitelisting is only checked during the Token Transfer, since we specifically don't enforce any rules in our ERC20 Token, whitelisting is not in effect. I'd say we should test with whitelisting after we finish all the functionality for users.